### DISCUSSION TITLE
Intersection point of two lines?

### DISCUSSION THREAD

**Discussion by ThomasCartier** [georust/geo#1190] [Q&A] [answered]
Hi,
I didn't find a way to calculate the intersection point of two lines in this crate, did I miss something ?

```
#[derive(Copy, Clone, Debug)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    pub fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }
}

#[derive(Copy, Clone, Debug)]
struct Line(Point, Point);

impl Line {
    pub fn intersect(self, other: Self) -> Option<Point> {
        let a1 = self.1.y - self.0.y;
        let b1 = self.0.x - self.1.x;
        let c1 = a1 * self.0.x + b1 * self.0.y;

        let a2 = other.1.y - other.0.y;
        let b2 = other.0.x - other.1.x;
        let c2 = a2 * other.0.x + b2 * other.0.y;

        let delta = a1 * b2 - a2 * b1;

        if delta == 0.0 {
            return None;
        }

        Some(Point {
            x: (b2 * c1 - b1 * c2) / delta,
            y: (a1 * c2 - a2 * c1) / delta,
        })
    }
}

fn main() {
    let l1 = Line(Point::new(0.0, 0.0), Point::new(1.0, 0.0));
    let l2 = Line(Point::new(2.0, 2.0), Point::new(2.0, 1.0));
    println!("{:?}", l1.intersect(l2));

}
```

That code is from Rosetta.
In this crate, we only know if they intersect, not their intersection point.
Interpolating with a value greater than 1.0 is saturated to 1.0. So no way ?

Thanks,

**Comment by michaelkirk** [comment]
See https://docs.rs/geo/latest/geo/algorithm/line_intersection/fn.line_intersection.html

**Reply by ThomasCartier** [reply]
Except:

```
let line_1 = Line::new(coord! {x: 0.0, y: 0.0}, coord! { x: 1.0, y: 0.0 });
let line_2 = Line::new(coord! {x: 2.0, y: 2.0}, coord! { x: 2.0, y: 1.0 });

println!("{:?}", line_intersection(line_1, line_2));
```

returns `None`, when it should return (2.0; 0.0), by extending the line.
The rosetta algo does the right thing in this regard.
Maybe an option could be passed as "extend_line: bool" ?

**Comment by urschrei** [comment]
The two lines in your example don't intersect:

![Screenshot 2024-06-05 at 17 51 16](https://github.com/georust/geo/assets/131862/10f5449e-32a6-48b6-9618-de76e6694c2b)

> by extending the line.

I'm not sure what you mean by this.

**Reply by ThomasCartier** [reply]
If you prolongate them, their point of intersection is indeed (2;0);
Like rays.

Is there a method in geo for this ?

**Reply by urschrei** [reply, marked as answer]
> Is there a method in geo for this ?

Not currently.

**Reply by ThomasCartier** [reply]
Thanks! just for clarity the image:
![336933127-10f5449e-32a6-48b6-9618-de76e6694c2b](https://github.com/georust/geo/assets/146366386/a2cbf1bf-1a82-4ef0-a1da-a57807277833)


Thanks urschrei
